python/ray/data/_internal/block_batching/iter_batches.py

```diff
@@ -1,47 +1,247 @@
+            manually (i.e. without waiting for Python's automatic GC) after it
+            is read. Doing so will reclaim memory faster and hence reduce the
+            memory footprint. However, the caller has to ensure the safety, i.e.
+            the block will never be accessed again.
+        batch_size: Record batch size, or None to let the system pick.
+        batch_format: The format in which to return each batch.
+            Specify "default" to use the current block format (promoting
+            Arrow to pandas automatically), "pandas" to
+            select ``pandas.DataFrame`` or "pyarrow" to select
+            ``pyarrow.Table``. Default is "default".
+        drop_last: Whether to drop the last batch if it's incomplete.
+        collate_fn: A function to apply to each data batch before returning it.
+        shuffle_buffer_min_size: If non-None, the data will be randomly shuffled using a
+            local in-memory shuffle buffer, and this value will serve as the minimum
+            number of rows that must be in the local in-memory shuffle buffer in order
+            to yield a batch.
+        shuffle_seed: The seed to use for the local random shuffle.
+        ensure_copy: Whether batches are always copied from the underlying base
+            blocks (not zero-copy views).
+        prefetch_batches: The number of batches to fetch ahead of the current batch to
+            process. If set to greater than 0, a separate thread will be used to fetch
+            the specified amount of formatted batches from blocks. This improves
+            performance for non-CPU bound UDFs, allowing batch fetching compute and
+            formatting to be overlapped with the UDF. Defaults to 0 (no prefetching
+            enabled).
+
+    Returns:
+        An iterator over record batches.
+    """
+    context = DatasetContext.get_current()
+
+    if (
+        prefetch_batches > 0
+        and context.actor_prefetcher_enabled
+        and not ray.util.client.ray.is_connected()
+    ):
+        prefetcher = ActorBlockPrefetcher()
+    else:
+        prefetcher = WaitBlockPrefetcher()
 
-If prefetch_batches=2, these are all the batches in flight:
+    eager_free = clear_block_after_read and DatasetContext.get_current().eager_free
 
-[User thread] trains on Batch 0
-   - [Fetch thread] Batch 1 in output queue
-        - [Worker thread 1] Batch 2 formatting + collating
-        - [Worker thread 2] Batch 3 formatting + collating
-        - [Raylet] Batches 4 + 5 fetched to local object store memory
+    def _async_iter_batches(
+        block_refs: Iterator[Tuple[ObjectRef[Block], BlockMetadata]],
+    ) -> Iterator[DataBatch]:
 
-At any point in time there are prefetch_batches+1 batches in local heap memory.
-And the next set of prefetch_batches in local object store memory.
+        if prefetch_batches > 0:
+            # Step 1: Prefetch logical batches locally.
+            block_refs = prefetch_batches_locally(
```

Comment:
Any way to always go through this code path (handle the zero case inside this function?) That would help avoid diverging code paths.

Reply:
updated